NEXT-12738 - Using the datepicker plugin (#234)
shopware/docs: guides/plugins/plugins/storefront/using-the-datepicker-plugin.md

# Using the datepicker plugin

## Overview

To provide an input field for date and time values, you can use the datepicker plugin. This guide shows you how to use it.

The datepicker plugin uses the `flatpickr` implementation under the hood. So, check out the `flatpickr` documentation,
if you need more information about the date picker configuration itself.

<!-- markdown-link-check-disable-next-line -->

> {% embed url="https://flatpickr.js.org" caption="" %}

Reviewer: Would you like to give a short caption or remove the attribute?

## Prerequisites

You won't learn how to create a plugin in this guide, head over to our Plugin base guide to create
your first plugin:

{% page-ref page="../plugin-base-guide.md" %}

You should also know how to customize templates:

{% page-ref page="./customize-templates.md" %}

## Setup a datepicker input field

> To apply the datepicker functionality we have to add a DOM element like an input field in a template.

Reviewer: ```suggestion
To apply the datepicker functionality we have to add a DOM element in a template, e.g. an input field.
```

To keep this example simple for now we just override the `base_main_inner` of the `storefront/page/content/index.html.twig`.

{% code title="<plugin root>/src/Resources/views/storefront/page/content/index.html.twig" %}

> ```text

Reviewer 2: Dunno, if this makes a difference, but we use twig at other places, so maybe we should stick to it

``````suggestion
```twig
``````

{% sw_extends '@Storefront/storefront/page/content/index.html.twig' %}

{% block base_main_inner %}
    <label>
        <input type="text"
               name="customDate"
               class="customDate"
        />
    </label>

    {{ parent() }}
{% endblock %}
```
{% endcode %}

Now you should see an empty input field if you open the storefront in your browser.

> We need to add the attribute data-attribute `data-date-picker` to activate the datepicker plugin on our input field.

Reviewer: This way?

```suggestion
We need to add the data-attribute `data-date-picker` to activate the datepicker plugin on our input field.
```

{% code title="<plugin root>/src/Resources/views/storefront/page/content/index.html.twig" %}

> ```text

Reviewer 2: ``````suggestion
```twig
``````

{% sw_extends '@Storefront/storefront/page/content/index.html.twig' %}

{% block base_main_inner %}
    <label>
        <input type="text"
               name="customDate"
               class="customDate"
               data-date-picker
        />
    </label>

    {{ parent() }}
{% endblock %}
```
{% endcode %}

> If we check the change in the browser again after reloading the page we now can see, that the datepicker plugin is now active on this element.

Reviewer: ```suggestion
If we check the change in the browser again, thus after reloading the page, we can see that the datepicker plugin is now active on this element.
```

## Configure the datepicker

If you select a date with the datepicker from the example above, you will see that a time is always selected and displayed in the input field. By default, the time selection is activated.

We can change this behaviour by passing more options to the datepicker plugin.

Here you can see how this is done by setting up a local Twig variable `pickerOptions`. We can assign a JSON formatted object to the variable and pass the value to the datepicker plugin through the `data-date-picker-options` attribute.

{% code title="<plugin root>/src/Resources/views/storefront/page/content/index.html.twig" %}

> ```text

Reviewer 2: ``````suggestion
```twig
``````

Reviewer 3: Rather HTML I guess? Don't think it can parse twig and will therefore default to `text` anyway.

HTML will at least parse something

{% sw_extends '@Storefront/storefront/page/content/index.html.twig' %}

{% block base_main_inner %}

>     {% set pickerOptions = '{

Reviewer 4: Are you building JSON here on your own? I suggest to use the `json_encode` filter before passing it into the `data-date-picker-options` attribute.

Reviewer 3: Just checked this in the core and it seems like we're always doing the "manual" way here. Not saying it's better, but at least more consistent with the core this way.

        "locale": "' ~ app.request.locale ~ '",
        "enableTime": true
    }' %}
    
    <label>
        <input type="text"
               name="customDate"
               class="customDate"
               data-date-picker
               data-date-picker-options="{{ pickerOptions }}"
        />
    </label>

    {{ parent() }}
{% endblock %}
```
{% endcode %}

> As you can see we also pass in the `locale` option which gets its value from `app.request.locale`. As a result,

Reviewer: ```suggestion
As you can see, we also pass in the `locale` option which gets its value from `app.request.locale`. As a result,
```

the datepicker plugin now uses the same locale as the current storefront and the date formatting matches active
languages accordingly.

## Preselect a date

> To preselect the value of the datepicker we simply can set it value in the input field which gets picked up by the datepicker plugin.

Reviewer: ```suggestion
To preselect the value of the datepicker we can simply set it value in the input field which gets picked up by the datepicker plugin.
```

Reviewer 2: ```suggestion
To preselect the value of the datepicker we can simply set its value in the input field which gets picked up by the datepicker plugin.
```

{% code title="<plugin root>/src/Resources/views/storefront/page/content/index.html.twig" %}

> ```text

Reviewer 2: ``````suggestion
```twig
``````

{% sw_extends '@Storefront/storefront/page/content/index.html.twig' %}

{% block base_main_inner %}

>     {% set pickerOptions = '{

Reviewer 4: Same JSON building issue

        "locale": "' ~ app.request.locale ~ '",
        "enableTime": true
    }' %}
    
    <label>
        <input type="text"
               name="customDate"
               class="customDate"
               value="2021-01-01T00:00:00+00:00"
               data-date-picker
               data-date-picker-options="{{ pickerOptions }}"
        />
    </label>

    {{ parent() }}
{% endblock %}
```
{% endcode %}

## Controlling the datepicker via buttons

To open or close the datepicker by trigger buttons you can pass in DOM selectors. You also can setup a selector to reset the currently selected value.
Here is an example which show all three selectors in action.

{% code title="<plugin root>/src/Resources/views/storefront/page/content/index.html.twig" %}

> ```text

Reviewer 2: ``````suggestion
```twig
``````

{% sw_extends '@Storefront/storefront/page/content/index.html.twig' %}

{% block base_main_inner %}

>     {% set pickerProperties = '{

Reviewer 4: Same JSON building issue

        "locale": "' ~ app.request.locale ~ '",
        "enableTime": true,
        "selectors": {
            "openButton" : ".openDatePicker",
            "closeButton": ".closeDatePicker",
            "clearButton": ".resetDatePicker"
        }
    }' %}

    <label>
        <input type="text"
               name="foo"
               class="customDate"
               value="2021-04-13T00:00:00+00:00"
               data-date-picker
               data-date-picker-options="{{ pickerProperties }}"
        />

        <button class="openDatePicker">Open</button>
        <button class="closeDatePicker">Close</button>
        <button class="resetDatePicker">Reset</button>
    </label>

    {{ parent() }}
{% endblock %}
```
{% endcode %}

## More options

| Option | Default | Description |
| :--- | :--- | :--- |

> | dateFormat | 'Y-m-dTH:i:S+00:00' | Pattern for the date string representation

Reviewer: Can you use `code syntax` for the options?
